torch_xla/csrc/aten_cpu_fallback.cpp

```diff
@@ -1,34 +1,355 @@
+  // Initialize CUDA device.
+  torch::utils::device_lazy_init(at::kCUDA);
+
+  // Step 1: Convert all non-CUDA tensor inputs into CUDA tensors
+  // and put them on the stack at the correct indices.
+  for (const auto idx : c10::irange(arguments.size())) {
+    const auto& ivalue = arguments[idx];
+    if (ivalue.isTensor()) {
+      tensor_args.push_back(ivalue.toTensor());
+      tensor_args_indices.push_back(idx);
+    } else if (ivalue.isTensorList()) {
+      // Note: we copy each TensorList argument to CUDA individually out of
+      // convenience, but XLA would benefit from materializing all tensor and
+      // TensorList args onto the CUDA at the same time. We can improve this if
+      // we need better perf for XLA's CUDA fallbacks.
+      tensorlist_args.push_back(ivalue.toTensorList());
+      tensorlist_args_indices.push_back(idx);
+      auto cuda_ivalue = c10::IValue(c10::List<at::Tensor>(
+          to_cuda(ivalue.toTensorList().vec(), &common_device)));
+      tensorlist_cuda_args.push_back(cuda_ivalue);
+      (*stack)[arguments_begin + idx] = std::move(cuda_ivalue);
+      tensorlist_args.push_back(ivalue.toTensorList());
+    } else if (ivalue.isOptionalTensorList()) {
+      auto opt_tensors = ivalue.toOptionalTensorList().vec();
+      std::vector<at::Tensor> need_convert_tensors;
+      std::vector<int> need_convert_tensors_index;
+      for (auto i : c10::irange(opt_tensors.size())) {
+        if (!opt_tensors[i].has_value() || !opt_tensors[i]->defined()) continue;
+        need_convert_tensors.push_back(opt_tensors[i].value());
+        need_convert_tensors_index.push_back(i);
+      }
+      auto cuda_tensors = to_cuda(need_convert_tensors, &common_device);
+      for (const auto i : c10::irange(need_convert_tensors_index.size())) {
+        auto idx = need_convert_tensors_index[i];
+        opt_tensors[idx] = cuda_tensors[i];
+      }
+      (*stack)[arguments_begin + idx] = c10::IValue(opt_tensors);
+    }
+  }
+  // XLA requires all of the tensor arguments to be gathered up and converted to
+  // CUDA together.
+  auto cuda_tensors = to_cuda(tensor_args, &common_device);
+
+  for (const auto i : c10::irange(tensor_args_indices.size())) {
+    auto idx = tensor_args_indices[i];
+    (*stack)[arguments_begin + idx] = c10::IValue(cuda_tensors[i]);
+  }
+
+  // Step 2: Call the underlying CUDA implementation of the operator
+  op.redispatchBoxed(c10::DispatchKeySet(c10::DispatchKey::CUDA), stack);
+
+  // Synchronize the device before actually converting back to XLA.
+  TORCH_CHECK(
+      common_device != -1,
+      "No defined tensors found for CUDA fallback: ", op.operator_name());
+  torch_cuda_synchronize(common_device);
+
+  // Step 3: We need to take special care to handle mutable aliases properly:
+  // If any input tensors are mutable aliases, we need to
+  // directly copy the updated data on the CUDA tensors back to the original
+  // inputs.
+  for (const auto i : c10::irange(tensor_args_indices.size())) {
```

Comment:
Does torch_xla has a concept of mutable aliases?
Also, do you know if there is a specific test cases for step3?

Reply:
It doesn't. The functional layer abstracts it for PyTorch/XLA. That's why we have to propagate the results back to the input arguments that were mutated.

Reply:
> Also, do you know if there is a specific test cases for step3?

Not sure we have those in PyTorch/XLA. For that, we would need an operation that mutates at least one of the inputs, e.g. operations that have the `out` parameter.